File: main/wallpaper.js
// ============================================================
// 主进程动态壁纸服务：桌面视频发现、GitHub 壁纸库列取与下载
// ============================================================
'use strict';

const path = require('path');
const fs = require('fs');
const http = require('http');
const https = require('https');

function streamDownload(url, dest, onProgress, redirects = 0) {
  return new Promise((resolve, reject) => {
    const lib = url.startsWith('https') ? https : http;
    const req = lib.get(url, { headers: { 'User-Agent': 'FuFumidi' } }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location && redirects < 5) {
        res.resume();
        const next = new URL(res.headers.location, url).toString();
        resolve(streamDownload(next, dest, onProgress, redirects + 1));
        return;
      }
      if (res.statusCode !== 200) { res.resume(); reject(new Error('HTTP ' + res.statusCode)); return; }
      const total = parseInt(res.headers['content-length'] || '0', 10) || 0;
      let received = 0;
      let lastPct = -1;
      const ws = fs.createWriteStream(dest);
      // 单 data 监听（不再同时 pipe，避免双监听）；进度按 ~2% 间隔节流，避免高频事件闪烁
      res.on('data', (chunk) => {
        received += chunk.length;
        const pct = total ? Math.round(received / total * 100) : 0;
        if (onProgress && total > 0 && (pct >= lastPct + 2 || pct === 100)) {
          lastPct = pct;
          try { onProgress(Math.max(0, Math.min(1, received / total))); } catch (e) {}
        }
        if (!ws.write(chunk)) {
          res.pause();
          ws.once('drain', () => res.resume());
        }
      });
      res.on('end', () => { ws.end(); });
      ws.on('finish', () => { try { if (onProgress) onProgress(1); } catch (e) {} resolve(); });
      ws.on('error', reject);
      res.on('error', reject);
    });
    req.on('error', reject);
  });
}

async function fetchThumbData(net, url) {
  try {
    const res = await net.fetch(url, { headers: { 'User-Agent': 'FuFumidi' } });
    if (!res.ok || !res.body) return '';
    const buf = Buffer.from(await res.arrayBuffer());
    const mime = /\.png$/i.test(url) ? 'image/png' : 'image/jpeg';
    return 'data:' + mime + ';base64,' + buf.toString('base64');
  } catch (e) { return ''; }
}

async function fetchThumbDataCached(net, url, cachePath) {
  try {
    if (fs.existsSync(cachePath)) {
      return 'data:image/jpeg;base64,' + fs.readFileSync(cachePath).toString('base64');
    }
    const data = await fetchThumbData(net, url);
    if (data && data.startsWith('data:image/')) {
      const base64 = data.split(',')[1];
      if (base64) {
        try { fs.writeFileSync(cachePath, Buffer.from(base64, 'base64')); } catch (e) {}
      }
    }
    return data;
  } catch (e) { return ''; }
}

function registerWallpaperIpc({ ipcMain, app, fs: f, net, runEngineInline, parsePyJson }) {
  // 动态壁纸：发现桌面上的视频文件（mp4/webm/mov），供渲染进程作为壁纸源
  ipcMain.handle('wallpaper:defaults', async () => {
    try {
      const desktop = app.getPath('desktop');
      const exts = new Set(['.mp4', '.webm', '.mov']);
      const out = [];
      if (f.existsSync(desktop)) {
        for (const file of f.readdirSync(desktop)) {
          const ext = path.extname(file).toLowerCase();
          if (exts.has(ext)) out.push(path.join(desktop, file));
        }
      }
      out.sort((a, b) => a.localeCompare(b, 'zh'));
      return { ok: true, files: out.slice(0, 4) };
    } catch (e) { return { ok: false, files: [] }; }
  });

  async function ensureLocalThumb(videoPath, dir) {
    try {
      const base = path.basename(videoPath).replace(/\.(mp4|webm|mov)$/i, '');
      const thumbPath = path.join(dir, base + '.jpg');
      if (f.existsSync(thumbPath)) {
        return 'data:image/jpeg;base64,' + f.readFileSync(thumbPath).toString('base64');
      }
      if (!runEngineInline) return '';
      const args = ['-y', '-hide_banner', '-loglevel', 'error', '-ss', '0.5', '-i', videoPath, '-frames:v', '1', '-vf', 'scale=320:-2', '-q:v', '4', thumbPath];
      const code =
        'import json, subprocess\n' +
        'import imageio_ffmpeg\n' +
        'ff = imageio_ffmpeg.get_ffmpeg_exe()\n' +
        'r = subprocess.run([ff] + ' + JSON.stringify(args) + ', capture_output=True)\n' +
        "print(json.dumps({'ok': r.returncode == 0, 'err': (r.stderr or b'').decode('utf-8', 'replace')[-300:]}))";
      const rr = await runEngineInline(code);
      const d = parsePyJson ? parsePyJson(rr.out) : null;
      if (d && d.ok && f.existsSync(thumbPath)) {
        return 'data:image/jpeg;base64,' + f.readFileSync(thumbPath).toString('base64');
      }
      return '';
    } catch (e) { return ''; }
  }

  async function listLocalWallpapers() {
    try {
      const dir = path.join(app.getPath('userData'), 'wallpapers');
      if (!f.existsSync(dir)) return [];
      const exts = new Set(['.mp4', '.webm', '.mov']);
      const candidates = [];
      for (const file of f.readdirSync(dir)) {
        const ext = path.extname(file).toLowerCase();
        if (!exts.has(ext)) continue;
        if (file.endsWith('.part')) continue;   // 下载中的临时文件，不算已下载
        const full = path.join(dir, file);
        // 过滤损坏/未完成的空文件（0 字节），避免列出假壁纸导致应用后黑屏
        try { if (f.statSync(full).size <= 0) continue; } catch (e) { continue; }
        // 只把「完整下载」的壁纸视为已下载（下载完成会写 .ok 标记）
        if (!f.existsSync(full + '.ok')) continue;
        candidates.push(full);
      }
      const thumbs = await Promise.all(candidates.map(videoPath => ensureLocalThumb(videoPath, dir)));
      const out = candidates.map((videoPath, i) => ({ name: path.basename(videoPath), video: videoPath, thumb: thumbs[i], local: true }));
      out.sort((a, b) => a.name.localeCompare(b.name, 'zh'));
      return out;
    } catch (e) { return []; }
  }

  // 动态壁纸库：从 GitHub Media 仓库列出壁纸（视频 + 同名缩略图），供用户选择下载
  const WALLPAPER_REPO = 'monologue82/Media';
  const WALLPAPER_DIR = 'wallpapers';
  const WALLPAPER_RAW = `https://raw.githubusercontent.com/${WALLPAPER_REPO}/main/${WALLPAPER_DIR}`;
  // Git LFS 视频需通过 media.githubusercontent.com 获取真实文件
  const WALLPAPER_MEDIA = `https://media.githubusercontent.com/media/${WALLPAPER_REPO}/main/${WALLPAPER_DIR}`;
  const thumbCacheDir = path.join(app.getPath('userData'), 'wallpaper-thumbs');
  try { f.mkdirSync(thumbCacheDir, { recursive: true }); } catch (e) {}
  // 远程壁纸目录缓存：5 分钟内直接复用 GitHub API 结果
  async function fetchRemoteList() {
    const cacheFile = path.join(thumbCacheDir, 'wallpaper-list.json');
    try {
      if (f.existsSync(cacheFile)) {
        const age = Date.now() - f.statSync(cacheFile).mtimeMs;
        if (age < 5 * 60 * 1000) {
          const cached = JSON.parse(f.readFileSync(cacheFile, 'utf8'));
          if (cached && Array.isArray(cached.vids) && Array.isArray(cached.thumbs)) return cached;
        }
      }
    } catch (e) {}
    const res = await net.fetch(`https://api.github.com/repos/${WALLPAPER_REPO}/contents/${WALLPAPER_DIR}`, {
      headers: { 'User-Agent': 'FuFumidi', Accept: 'application/vnd.github+json' },
    });
    if (!res.ok) throw new Error('GitHub API ' + res.status);
    const items = await res.json();
    const vids = items.filter(x => x.type === 'file' && /\.(mp4|webm|mov)$/i.test(x.name));
    const thumbs = items.filter(x => x.type === 'file' && /\.(jpg|jpeg|png)$/i.test(x.name));
    try { f.writeFileSync(cacheFile, JSON.stringify({ vids, thumbs })); } catch (e) {}
    return { vids, thumbs };
  }
  ipcMain.handle('wallpaper:list', async () => {
    try {
      const { vids, thumbs } = await fetchRemoteList();
      const remoteThumbs = await Promise.all(vids.map(async (f) => {
        const base = f.name.replace(/\.(mp4|webm|mov)$/i, '');
        const th = thumbs.find(t => t.name.replace(/\.(jpg|jpeg|png)$/i, '') === base);
        let thumb = th ? `${WALLPAPER_RAW}/${encodeURIComponent(th.name)}` : '';
        if (thumb) {
          const safeName = (th.name || f.name).replace(/[\\/:*?"<>|]/g, '_');
          thumb = await fetchThumbDataCached(net, thumb, path.join(thumbCacheDir, safeName));
        }
        return { name: f.name, video: `${WALLPAPER_MEDIA}/${encodeURIComponent(f.name)}`, remote: `${WALLPAPER_MEDIA}/${encodeURIComponent(f.name)}`, thumb };
      }));
      // 合并本地壁纸：远程项若本地已有同名文件 → 标记为已下载（local），不重复新增项
      const locals = await listLocalWallpapers();
      const localByBase = new Map();
      for (const l of locals) localByBase.set(l.name.replace(/\.(mp4|webm|mov)$/i, '').toLowerCase(), l);
      const list = remoteThumbs.map((r) => {
        const base = r.name.replace(/\.(mp4|webm|mov)$/i, '').toLowerCase();
        const loc = localByBase.get(base);
        if (loc) return { ...r, local: true, video: loc.video, thumb: loc.thumb || r.thumb };
        return r;
      });
      // 仅本地存在的壁纸（用户自行导入）也展示
      const seen = new Set(list.map(r => r.name.replace(/\.(mp4|webm|mov)$/i, '').toLowerCase()));
      const extra = locals.filter(l => !seen.has(l.name.replace(/\.(mp4|webm|mov)$/i, '').toLowerCase()));
      return { ok: true, list: list.concat(extra) };
    } catch (e) { return { ok: false, error: String((e && e.message) || e) }; }
  });

  // 删除本地壁纸：移除视频与对应缩略图
  ipcMain.handle('wallpaper:removeLocal', async (_e, nameOrPath) => {
    try {
      const name = path.basename(String(nameOrPath || ''));
      if (!name) return { ok: false, error: 'invalid name' };
      const dir = path.join(app.getPath('userData'), 'wallpapers');
      const video = path.resolve(dir, name);
      if (!video.startsWith(path.resolve(dir) + path.sep)) return { ok: false, error: 'path outside wallpaper dir' };
      if (!f.existsSync(video)) return { ok: false, error: 'not found' };
      f.unlinkSync(video);
      try { f.unlinkSync(video + '.ok'); } catch (e) {}   // 完成标记一并删除
      const thumb = path.join(dir, name.replace(/\.(mp4|webm|mov)$/i, '') + '.jpg');
      if (f.existsSync(thumb)) f.unlinkSync(thumb);
      return { ok: true };
    } catch (e) { return { ok: false, error: String((e && e.message) || e) }; }
  });

  // 本地壁纸导入：将用户选择的视频复制到 userData/wallpapers，加入壁纸库
  ipcMain.handle('wallpaper:addLocal', async (event, srcPath) => {
    try {
      if (!srcPath || typeof srcPath !== 'string' || !f.existsSync(srcPath)) return { ok: false, error: '本地文件不存在' };
      const ext = path.extname(srcPath).toLowerCase();
      if (!['.mp4', '.webm', '.mov'].includes(ext)) return { ok: false, error: '不支持的视频格式' };
      const dir = path.join(app.getPath('userData'), 'wallpapers');
      f.mkdirSync(dir, { recursive: true });
      const base = path.basename(srcPath).replace(/[\/:*?"<>|]/g, '_') || 'local-wallpaper.mp4';
      let dest = path.join(dir, base);
      let n = 1;
      const extname = path.extname(base);
      const stem = extname ? base.slice(0, -extname.length) : base;
      while (f.existsSync(dest) && path.resolve(dest) !== path.resolve(srcPath)) {
        dest = path.join(dir, stem + '-' + n + extname);
        n += 1;
      }
      const totalSize = f.statSync(srcPath).size || 0;
      let copied = 0;
      const notify = (p) => {
        try { if (event && !event.sender.isDestroyed()) event.sender.send('wallpaper:addLocalProgress', { path: dest, progress: Math.max(0, Math.min(1, p)) }); } catch (e) {}
      };
      notify(0);
      await new Promise((resolve, reject) => {
        const rs = f.createReadStream(srcPath);
        const ws = f.createWriteStream(dest);
        rs.on('data', (chunk) => {
          copied += chunk.length;
          if (totalSize > 0) notify(copied / totalSize);
        });
        ws.on('finish', resolve);
        ws.on('error', reject);
        rs.on('error', reject);
        rs.pipe(ws);
      });
      notify(1);
      return { ok: true, path: dest, name: path.basename(dest) };
    } catch (e) { return { ok: false, error: String((e && e.message) || e) }; }
  });

  // 动态壁纸下载：流式下载视频到 userData/wallpapers，返回本地路径（带进度 + 完整性校验）
  // 先写 .part 临时文件，下载完整后再改名，避免中途退出残留被当作"已下载"
  ipcMain.handle('wallpaper:download', async (evt, url, name) => {
    const dir = path.join(app.getPath('userData'), 'wallpapers');
    try {
      if (!url || !/^https?:\/\//i.test(url)) return { ok: false, error: '无效 URL' };
      f.mkdirSync(dir, { recursive: true });
      const safe = String(name || 'wallpaper').replace(/[\\/:*?"<>|]/g, '_');
      const dest = path.join(dir, safe);
      const tmp = dest + '.part';
      const sendP = (p) => {
        try { if (evt && !evt.sender.isDestroyed()) evt.sender.send('wallpaper:downloadProgress', { name: safe, progress: p }); } catch (e) {}
      };
      sendP(0);
      await streamDownload(url, tmp, (p) => sendP(p));
      // 完整性校验：0 字节 = 下载失败/被拦截，删除残留避免假壁纸
      let size = 0;
      try { size = f.statSync(tmp).size; } catch (e) {}
      if (size <= 0) {
        try { f.unlinkSync(tmp); } catch (e) {}
        return { ok: false, error: '下载文件为空（网络受限或文件被拦截），已回退在线壁纸' };
      }
      // 改名到正式文件名（原子替换：先删旧再改名），并写完成标记 .ok
      if (f.existsSync(dest)) { try { f.unlinkSync(dest); } catch (e) {} }
      f.renameSync(tmp, dest);
      try { f.writeFileSync(dest + '.ok', '1', 'utf8'); } catch (e) {}
      sendP(1);
      return { ok: true, path: dest, name: safe, size };
    } catch (e) {
      const safe = String(name || 'wallpaper').replace(/[\\/:*?"<>|]/g, '_');
      const tmp = path.join(dir, safe + '.part');
      try { if (f.existsSync(tmp)) f.unlinkSync(tmp); } catch (e2) {}
      return { ok: false, error: String((e && e.message) || e) };
    }
  });
}

module.exports = { registerWallpaperIpc };

File: src-tauri/src/wallpaper.rs
// 动态壁纸服务：桌面视频发现 / GitHub 壁纸库 / 本地导入删除 / 下载 —— 移植自 main/wallpaper.js
use serde_json::{json, Value};
use std::fs;
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Emitter, Manager};

const WALLPAPER_REPO: &str = "monologue82/Media";
const WALLPAPER_DIR: &str = "wallpapers";
const WALLPAPER_API: &str = "https://api.github.com/repos/monologue82/Media/contents/wallpapers";
const WALLPAPER_MEDIA: &str = "https://media.githubusercontent.com/media/monologue82/Media/main/wallpapers";
const WALLPAPER_RAW: &str = "https://raw.githubusercontent.com/monologue82/Media/main/wallpapers";

fn http_get(url: &str) -> Result<String, String> {
    let out = std::process::Command::new("curl.exe")
        .args(["--ssl-no-revoke", "-sS", "-L", "-m", "20", "-H", "User-Agent: FuFumidi", url])
        .output()
        .map_err(|e| e.to_string())?;
    if out.status.success() {
        Ok(String::from_utf8_lossy(&out.stdout).to_string())
    } else {
        Err(String::from_utf8_lossy(&out.stderr).to_string())
    }
}

fn http_download(url: &str, dest: &Path) -> Result<(), String> {
    let out = std::process::Command::new("curl.exe")
        .args(["--ssl-no-revoke", "-sS", "-L", "-o"])
        .arg(dest.to_string_lossy().to_string())
        .arg(url)
        .output()
        .map_err(|e| e.to_string())?;
    if out.status.success() {
        Ok(())
    } else {
        Err(String::from_utf8_lossy(&out.stderr).to_string())
    }
}

fn wallpaper_dir(app: &AppHandle) -> PathBuf {
    app.path()
        .app_data_dir()
        .unwrap_or_else(|_| PathBuf::from("."))
        .join("wallpapers")
}

fn is_video(name: &str) -> bool {
    let n = name.to_lowercase();
    n.ends_with(".mp4") || n.ends_with(".webm") || n.ends_with(".mov")
}

// 发现桌面上的视频文件（最多 4 个）
#[tauri::command]
pub fn wallpaper_defaults(app: AppHandle) -> Value {
    let mut out: Vec<String> = Vec::new();
    if let Ok(desktop) = app.path().desktop_dir() {
        if let Ok(entries) = fs::read_dir(&desktop) {
            let mut files: Vec<String> = entries
                .flatten()
                .filter(|e| {
                    let name = e.file_name().to_string_lossy().to_string();
                    is_video(&name)
                })
                .map(|e| e.path().to_string_lossy().to_string())
                .collect();
            files.sort_by(|a, b| a.to_lowercase().cmp(&b.to_lowercase()));
            out = files.into_iter().take(4).collect();
        }
    }
    json!({ "ok": true, "files": out })
}

// 本地已下载壁纸列表
fn list_local(app: &AppHandle) -> Vec<Value> {
    let dir = wallpaper_dir(app);
    let mut out = Vec::new();
    if let Ok(entries) = fs::read_dir(&dir) {
        for e in entries.flatten() {
            let p = e.path();
            let name = p.file_name().and_then(|n| n.to_str()).unwrap_or("");
            // 只列「完整下载」的壁纸：视频 + .ok 标记 + 非 0 字节
            if !is_video(name) || name.ends_with(".part") {
                continue;
            }
            if !p.with_extension("ok").exists() {
                continue;
            }
            let size = fs::metadata(&p).map(|m| m.len()).unwrap_or(0);
            if size <= 0 {
                continue;
            }
            out.push(json!({ "name": name, "video": p.to_string_lossy(), "thumb": "", "local": true }));
        }
    }
    out.sort_by(|a, b| a["name"].as_str().unwrap_or("").cmp(b["name"].as_str().unwrap_or("")));
    out
}

// 壁纸库列表（GitHub 远程 + 本地合并）
#[tauri::command]
pub fn wallpaper_list(app: AppHandle) -> Value {
    let body = match http_get(WALLPAPER_API) {
        Ok(b) => b,
        Err(e) => return json!({ "ok": false, "error": e }),
    };
    let items: Vec<Value> = match serde_json::from_str(&body) {
        Ok(v) => v,
        Err(_) => return json!({ "ok": false, "error": "GitHub API 解析失败" }),
    };
    let mut vids: Vec<Value> = Vec::new();
    let mut thumbs: Vec<Value> = Vec::new();
    for it in &items {
        if let Some(n) = it["name"].as_str() {
            if it["type"].as_str() == Some("file") {
                if is_video(n) {
                    vids.push(it.clone());
                } else if n.ends_with(".jpg") || n.ends_with(".jpeg") || n.ends_with(".png") {
                    thumbs.push(it.clone());
                }
            }
        }
    }
    let mut list: Vec<Value> = Vec::new();
    for v in &vids {
        let name = v["name"].as_str().unwrap_or("").to_string();
        let base = name.trim_end_matches(".mp4").trim_end_matches(".webm").trim_end_matches(".mov");
        let mut thumb = String::new();
        for t in &thumbs {
            let tn = t["name"].as_str().unwrap_or("");
            let tb = tn.trim_end_matches(".jpg").trim_end_matches(".jpeg").trim_end_matches(".png");
            if tb == base {
                thumb = format!("{}/{}", WALLPAPER_RAW, tn);
                break;
            }
        }
        list.push(json!({
            "name": name,
            "video": format!("{}/{}", WALLPAPER_MEDIA, name),
            "remote": format!("{}/{}", WALLPAPER_MEDIA, name),
            "thumb": thumb,
        }));
    }
    // 合并本地
    let locals = list_local(&app);
    let local_names: Vec<String> = locals.iter().filter_map(|l| l["name"].as_str().map(|s| s.to_lowercase())).collect();
    for l in &locals {
        let ln = l["name"].as_str().unwrap_or("").to_lowercase();
        if let Some(r) = list.iter_mut().find(|r| {
            r["name"].as_str().unwrap_or("").to_lowercase().replace(".mp4", "").replace(".webm", "").replace(".mov", "") == ln.replace(".mp4", "").replace(".webm", "").replace(".mov", "")
        }) {
            *r = l.clone();
        } else if !local_names.iter().any(|n| {
            n.replace(".mp4", "").replace(".webm", "").replace(".mov", "") == ln.replace(".mp4", "").replace(".webm", "").replace(".mov", "") && n != &ln
        }) {
            // 本地独有（自行导入）
            list.push(l.clone());
        }
    }
    json!({ "ok": true, "list": list })
}

// 删除本地壁纸
#[tauri::command]
pub fn wallpaper_remove_local(app: AppHandle, name: String) -> Value {
    let dir = wallpaper_dir(&app);
    let base = Path::new(&name)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("")
        .to_string();
    if base.is_empty() {
        return json!({ "ok": false, "error": "invalid name" });
    }
    let video = dir.join(&base);
    if !video.exists() {
        return json!({ "ok": false, "error": "not found" });
    }
    let _ = fs::remove_file(&video);
    let _ = fs::remove_file(dir.join(format!("{}.ok", base)));
    let stem = base.trim_end_matches(".mp4").trim_end_matches(".webm").trim_end_matches(".mov");
    let _ = fs::remove_file(dir.join(format!("{}.jpg", stem)));
    json!({ "ok": true })
}

// 本地壁纸导入：复制到 userData/wallpapers
#[tauri::command]
pub async fn wallpaper_add_local(app: AppHandle, src_path: String) -> Value {
    let src = PathBuf::from(&src_path);
    if !src.exists() || !src.is_file() {
        return json!({ "ok": false, "error": "本地文件不存在" });
    }
    let ext = src.extension().and_then(|e| e.to_str()).unwrap_or("").to_lowercase();
    if !["mp4", "webm", "mov"].contains(&ext.as_str()) {
        return json!({ "ok": false, "error": "不支持的视频格式" });
    }
    let dir = wallpaper_dir(&app);
    let _ = fs::create_dir_all(&dir);
    let raw_name = src.file_name().and_then(|n| n.to_str()).unwrap_or("local-wallpaper.mp4").to_string();
    let safe = raw_name.replace(['/', '\\', ':', '*', '?', '"', '<', '>', '|'], "_");
    let mut dest = dir.join(&safe);
    let mut n = 1usize;
    let (stem, extname) = match safe.rfind('.') {
        Some(i) => (safe[..i].to_string(), safe[i..].to_string()),
        None => (safe.clone(), String::new()),
    };
    while dest.exists() && dest != src {
        dest = dir.join(format!("{}-{}{}", stem, n, extname));
        n += 1;
    }
    let total = fs::metadata(&src).map(|m| m.len()).unwrap_or(0);
    let _ = app.emit("wallpaper:addLocalProgress", json!({ "path": dest.to_string_lossy(), "progress": 0 }));
    match fs::copy(&src, &dest) {
        Ok(_) => {
            let _ = app.emit("wallpaper:addLocalProgress", json!({ "path": dest.to_string_lossy(), "progress": 1 }));
            let _ = fs::write(dest.with_extension("ok"), "1");
            json!({ "ok": true, "path": dest.to_string_lossy(), "name": dest.file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default(), "size": total })
        }
        Err(e) => json!({ "ok": false, "error": e.to_string() }),
    }
}

// 壁纸下载：curl 下载 → 改名 + .ok 标记
#[tauri::command]
pub async fn wallpaper_download(app: AppHandle, url: String, name: Option<String>) -> Value {
    if !url.starts_with("http://") && !url.starts_with("https://") {
        return json!({ "ok": false, "error": "无效 URL" });
    }
    let dir = wallpaper_dir(&app);
    let _ = fs::create_dir_all(&dir);
    let raw = name.unwrap_or_else(|| "wallpaper.mp4".into());
    let safe = raw.replace(['/', '\\', ':', '*', '?', '"', '<', '>', '|'], "_");
    let dest = dir.join(&safe);
    let tmp = dir.join(format!("{}.part", safe));
    let _ = app.emit("wallpaper:downloadProgress", json!({ "name": safe, "progress": 0 }));
    match http_download(&url, &tmp) {
        Ok(()) => {
            let size = fs::metadata(&tmp).map(|m| m.len()).unwrap_or(0);
            if size <= 0 {
                let _ = fs::remove_file(&tmp);
                return json!({ "ok": false, "error": "下载文件为空（网络受限或文件被拦截），已回退在线壁纸" });
            }
            if dest.exists() {
                let _ = fs::remove_file(&dest);
            }
            if fs::rename(&tmp, &dest).is_ok() {
                let _ = fs::write(dest.with_extension("ok"), "1");
                let _ = app.emit("wallpaper:downloadProgress", json!({ "name": safe, "progress": 1 }));
                json!({ "ok": true, "path": dest.to_string_lossy(), "name": safe, "size": size })
            } else {
                json!({ "ok": false, "error": "文件改名失败" })
            }
        }
        Err(e) => {
            let _ = fs::remove_file(&tmp);
            json!({ "ok": false, "error": e })
        }
    }
}
